index.js

```diff
@@ -1,174 +1,221 @@
-      opts = Object.assign({}, opts)
-      Object.defineProperty(this, 'opts', readonly(opts))
+    if (typeof opts.source === 'undefined' &&
+        typeof opts.data === 'undefined') {
+      throw new Error('Block instances must be created with either an encode source or data')
     }
-
-    source () {
-      if (this.opts.cid || this.opts.data ||
-          this._encoded || this._decoded) return null
-      if (!this.opts.source) return null
-      return this.opts.source
+    if (typeof opts.source !== 'undefined' && !opts.codec && !opts.code) {
+      throw new Error('Block instances created from source objects must include desired codec')
     }
-
-    async cid () {
-      if (this.opts.cid) return this.opts.cid
-      const hash = await multihash.hash(this.encodeUnsafe(), this.opts.algo)
-      const cid = CID.create(1, this.code, hash)
-      this.opts.cid = cid
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return cid
+    if (opts.data && !opts.cid && !opts.codec && !opts.code) {
+      throw new Error('Block instances created from data must include cid or codec')
     }
+    opts.hasher = opts.hasher || sha256
+    // Do our best to avoid accidental mutations of the options object after instantiation
+    // Note: we can't actually freeze the object because we mutate it once per property later
+    opts = Object.assign({}, opts)
+    Object.defineProperty(this, 'opts', readonly(opts))
+    Object.defineProperty(this, 'asBlock', readonly(this))
+  }
 
-    get codec () {
-      if (this.opts.code) {
-        this.opts.codec = multicodec.get(this.opts.code).name
+  get hasher () {
+    if (this.opts.cid) {
+      if (!this.opts.hasher) {
+        this.opts.hasher = Block.codecs.get(this.opts.cid.multihash.code)
+      } else if (this.opts.hasher.code !== this.opts.cid.multihash.code) {
+        this.opts.hasher = Block.codecs.get(this.opts.cid.multihash.code)
+      } else {
+        return this.opts.hasher || sha256
       }
-      if (this.opts.cid) {
-        this.opts.codec = multicodec.get(this.opts.cid.code).name
-      }
-      return this.opts.codec
+      if (!this.opts.hasher) throw new Error('Do not have hash implementation')
     }
+    return this.opts.hasher || sha256
+  }
+
+  source () {
+    if (this.opts.cid || this.opts.data ||
+        this._encoded || this._decoded) return null
+    if (!this.opts.source) return null
+    return this.opts.source
+  }
+
+  async cid () {
+    if (this.opts.cid) return this.opts.cid
+    const hash = await this.hasher.digest(this.encodeUnsafe())
+    const cid = CID.create(1, this.opts.codec.code, hash)
+    this.opts.cid = cid
+    // https://github.com/bcoe/c8/issues/135
+    /* c8 ignore next */
+    return cid
+  }
 
-    get code () {
-      if (this.opts.cid) return this.opts.cid.code
-      if (!this.opts.code) {
-        this.opts.code = multicodec.get(this.codec).code
+  get codec () {
```

Comment:
❓Why not return codec from here just like `hasher` returns `hasher` ? User can lookup `.name` if that's what the y need.

Reply:
I didn’t want to take that breaking change yet because it’s going to cause a lot of churn in the tests. I also hate to change the value of existing properties rather than just migrating to new properties because the errors during migration tend to be much more painful, but maybe it’s worth it here since `codec` is really the only good name for this.